{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Основные задания (17 баллов)\n",
    "\n",
    "1. Напишите простой декоратор, подменивающий возвращаемое значение декорируемой функции на время её выполнения (Example_1). Для измерения времени воспользуйтесь модулем time. (3 балла)\n",
    "\n",
    "\n",
    "2. Напишите декоратор, позволяющий логировать запуски функций, распечатывая входные данные и тип возвращаемых значений (Example_2). Для получения имени класса в виде строки можно воспользоваться атрибутом __name__. (7 баллов)\n",
    "\n",
    "\n",
    "3. Сделайте декоратор - русскую рулетку, который сделает так, чтобы декорируемая функция с заданной вероятностью заменяла возвращаемое значение на переданное декоратору (Example_3). (7 баллов)\n",
    "\n",
    "Дополнительные задания (21 балл)\n",
    "\n",
    "\n",
    "1. Реализуйте декоратор аналогичный staticmethod без использования любых модулей (7 баллов)\n",
    "2. Реализуйте декоратор аналогичный dataclass.  Весь функционал сделать как у стандартного dataclass за исключением того, что он не должен принимать дополнительные аргументы. При работе с декораторами классов очень полезными могут оказаться встроенные в питон функции setattr и getattr. Также вам нужно будет найти способ \"распарсить\" аннотации типов. (14 баллов)\n",
    "\n",
    "Общие требования:\n",
    "Наличие README с описанием и инструкцией по использованию поощрается\n",
    "Сторонние библиотеки указываем в requirements.txt (но тут можно обойтись и без них)\n",
    "Форма сдачи - ссылка на пулл реквест в master/main ветку вашего репозитория BI_2021_Python\n",
    "Решение оформить в виде Python скрипта\n",
    "В скрипте должен быть пример использования (можно сделать его в блоке if __name__ == \"__main__\")\n",
    "Ваши декораторы должны работать с любыми функциями независимо от их аргументов и возвращаемых значений"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Wrapper that replaces function return with time of execution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.4543533325195312e-05"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def timing(func):\n",
    "    \"\"\"Decorator that replaces any fucntion return with function's execution time\n",
    "    :param function func: function to be wrapped\n",
    "    :returns:  wrapped function\"\"\"\n",
    "    def inner_func(*args, **kwargs):\n",
    "        start = time.time()\n",
    "        func(*args, **kwargs)\n",
    "        return time.time() - start\n",
    "    return inner_func\n",
    "\n",
    "\n",
    "@timing\n",
    "def power(a, b):\n",
    "    return a**b\n",
    "\n",
    "power(10,b=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Logging wrapper.  \n",
    "prints args, kwargs and type of function's return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Function power2 was called with positional arguments:\\ (10,) and keyword arguments: {'b': 11}\n",
      "Function power2 return has type list\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[10000000000, 285311670611]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def function_logging(func):\n",
    "    \"\"\"Wrapper that prints input args/kwargs and type of funtions return\n",
    "    :param function func: function to be wrapped\n",
    "    :returns:  wrapped function\n",
    "    \"\"\"\n",
    "    def inner_func(*args, **kwargs):\n",
    "        kv_string = \", \".join([f'{k}={v}'for k, v in kwargs.items()])\n",
    "        if not args and not kwargs:\n",
    "            print(f\"Function {func.__name__} was called without arguments\")\n",
    "        elif args and kwargs:\n",
    "            print(f\"Function {func.__name__} was called with positional arguments:\\ {args} and keyword arguments: {kwargs}\")\n",
    "        elif args:\n",
    "            print(f\"Function {func.__name__} was called with positional arguments: {args}\")\n",
    "        else:\n",
    "            print(f\"Function {func.__name__} was called with keyword arguments: {kv_string}\")\n",
    "        result = func(*args, **kwargs)    \n",
    "        print(f\"Function {func.__name__} return has type {type(result).__name__}\")\n",
    "        return result\n",
    "    return inner_func\n",
    "\n",
    "@function_logging\n",
    "def power2(a=5, b=5):\n",
    "    return [a**a, b**b]\n",
    "\n",
    "power2(10, b=11)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Wrapper randomly replacing function's return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z\n",
      "Z\n",
      "Z\n",
      "Oops, I did it again, I played with your heart, got lost in the gameZ\n",
      "Oops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameZ\n",
      "ZZZZZ\n",
      "Oops, I did it again, I played with your heart, got lost in the gameZZZ\n",
      "Oops, I did it again, I played with your heart, got lost in the game\n",
      "ZOops, I did it again, I played with your heart, got lost in the gameZZZZZZOops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameZZZZOops, I did it again, I played with your heart, got lost in the gameZOops, I did it again, I played with your heart, got lost in the gameZ\n",
      "Oops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameZZOops, I did it again, I played with your heart, got lost in the gameZZZOops, I did it again, I played with your heart, got lost in the gameOops, I did it again, I played with your heart, got lost in the gameZ\n"
     ]
    }
   ],
   "source": [
    "def rus_roulette(probability=0.5, return_value='Damm it! Try again'):\n",
    "    \"\"\"This wrapper with a given probability replaces output with retun_value\n",
    "    :param probability: int, float - sets probability of replacement\n",
    "    :param return_value: any - alternative return that replaces true one\n",
    "    :returns: wrapped function\n",
    "    \"\"\"\n",
    "    def wrapper_itself(func):\n",
    "        def inner_function(*args, **kwargs):\n",
    "            if random.random() < probability:\n",
    "                return return_value\n",
    "            return func(*args, **kwargs)\n",
    "        return inner_function\n",
    "    return wrapper_itself\n",
    "\n",
    "@rus_roulette(0.1, \"Oops, I did it again, I played with your heart, got lost in the game\")\n",
    "def snore(n):\n",
    "    return snore(n - 1) + snore(n - 2) if n > 2 else 'Z'\n",
    "\n",
    "for i in range(10):\n",
    "    print(snore(i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. MyStaticMethod\n",
    "Wrapper that replicates original staticmethod functionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not static\n",
      "fffffffffffffffffffffffffffffffffffffff\n",
      "fff\n",
      "static\n",
      "fff\n",
      "fff\n",
      "My_static\n",
      "fff\n",
      "fff\n"
     ]
    }
   ],
   "source": [
    "class MyStaticMethod:\n",
    "    \"\"\"This wrapper allows to call function directly from class lile ausual function .\n",
    "    Basically, identical to implementation from Python documentaion.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, func):\n",
    "        self.func = func\n",
    "\n",
    "    def __get__(self, obj, objtype=None):\n",
    "        return self.func\n",
    "\n",
    "    def __call__(self, *args, **kwds):\n",
    "        return self.f(*args, **kwds)\n",
    "    \n",
    "    \n",
    "class MyClass:\n",
    "\n",
    "    def __init__(self, num=5, num2=20):\n",
    "        self.num = num\n",
    "        self.num2 = num2\n",
    "\n",
    "    @MyStaticMethod \n",
    "    def respect_mystatic(num_f=39):\n",
    "        return \"f\"*num_f\n",
    "    \n",
    "    @staticmethod \n",
    "    def respect_static(num_f=39):\n",
    "        return \"f\"*num_f\n",
    "\n",
    "    def respect(self, num_f=39):\n",
    "        return \"f\"*num_f\n",
    "\n",
    "\n",
    "print(\"not static\")\n",
    "print(MyClass.respect(3))\n",
    "print(MyClass().respect(3))\n",
    "print(\"static\")\n",
    "print(MyClass.respect_static(3))\n",
    "print(MyClass().respect_static(3))\n",
    "print(\"My_static\")\n",
    "print(MyClass.respect_mystatic(3))\n",
    "print(MyClass().respect_mystatic(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. adjust class wrapper\n",
    "Imagine that we have installed a certain library that contains two classes `MyString` and `MySet`, which are inheritors of `str` and `set`, but also carry additional methods.\n",
    "\n",
    "The problem is that the library was written by not very careful people, so it turned out that some methods do not return the data type that we expect. For example, `MyString().reverse()` returns an object of class `str`, although it would be more logical to expect an object of class `MyString`.\n",
    "\n",
    "Find and implement a convenient way to make such methods return an instance of the current class, not the parent. In this case, **method code cannot be changed**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '__main__.MyString'>\n",
      "<class '__main__.MyString'>\n",
      "<class '__main__.MyString'>\n",
      "<class '__main__.MyString'>\n",
      "\n",
      "<class 'bool'>\n",
      "<class 'bool'>\n",
      "<class '__main__.MySet'>\n",
      "<class '__main__.MySet'>\n",
      "Da Bb Cc\n",
      "<class '__main__.MyString'>\n"
     ]
    }
   ],
   "source": [
    "def type_for_methods(cls):\n",
    "    '''Searches for public class methods and sets type for their output\n",
    "    :param class cls: class\n",
    "    :returns: class with new type for class methods outputs\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    for attr in dir(cls):\n",
    "        if not attr.startswith('_') and callable(getattr(cls, attr)):\n",
    "            setattr(cls, attr, change_type(getattr(cls, attr), cls))\n",
    "    return cls\n",
    "\n",
    "\n",
    "def change_type(func, cls):\n",
    "    '''Function change_type() is a decorator for class and changes type of class method output \n",
    "    with MyString for strings and MySet for sets\n",
    "    \n",
    "    Parameters:\n",
    "    :param func: class method\n",
    "    :param cls: class\n",
    "    :returns: inner_function\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    def inner_function(*args, **kwargs):\n",
    "        '''Determines the type of class method output and sets appropriate one\n",
    "        \n",
    "        :param args: arguments of funtion\n",
    "        :param kwargs: keyword arguments of function\n",
    "        :returns: result (MyString or MySet object)\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        result = func(*args, **kwargs)\n",
    "        if (cls == MyString and isinstance(result, str)) or ( # isinstance to strict str or set type output\n",
    "            cls == MySet and isinstance(result, set)):\n",
    "            return cls(result)\n",
    "        else:\n",
    "            return result\n",
    "    return inner_function\n",
    "\n",
    "\n",
    "@type_for_methods\n",
    "class MyString(str):\n",
    "    def reverse(self):\n",
    "        return self.__class__(self[::-1])\n",
    "    \n",
    "    def make_uppercase(self):\n",
    "        return \"\".join([chr(ord(char) - 32) if 97 <= ord(char) <= 122 else char for char in self])\n",
    "    \n",
    "    def make_lowercase(self):\n",
    "        return \"\".join([chr(ord(char) + 32) if 65 <= ord(char) <= 90 else char for char in self])\n",
    "    \n",
    "    def capitalize_words(self):\n",
    "        return \" \".join([word.capitalize() for word in self.split()])\n",
    "\n",
    "    \n",
    "@type_for_methods    \n",
    "class MySet(set):\n",
    "    def is_empty(self):\n",
    "        return len(self) == 0\n",
    "    \n",
    "    def has_duplicates(self):\n",
    "        return len(self) != len(set(self))\n",
    "    \n",
    "    def union_with(self, other):\n",
    "        return self.union(other)\n",
    "    \n",
    "    def intersection_with(self, other):\n",
    "        return self.intersection(other)\n",
    "    \n",
    "    def difference_with(self, other):\n",
    "        return self.difference(other)\n",
    "    \n",
    "string_example = MyString(\"Aa Bb Cc\")\n",
    "set_example_1 = MySet({1, 2, 3, 4})\n",
    "set_example_2 = MySet({3, 4, 5, 6, 6})\n",
    "\n",
    "print(type(string_example.reverse()))\n",
    "print(type(string_example.make_uppercase()))\n",
    "print(type(string_example.make_lowercase()))\n",
    "print(type(string_example.capitalize_words()))\n",
    "print()\n",
    "print(type(set_example_1.is_empty()))\n",
    "print(type(set_example_2.has_duplicates()))\n",
    "print(type(set_example_1.union_with(set_example_2)))\n",
    "print(type(set_example_1.difference_with(set_example_2)))\n",
    "print(string_example.replace(\"A\", \"D\"))\n",
    "print(type(string_example.replace(\"A\", \"D\")))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. OpenFasta Context Manager\n",
    "\n",
    "Context managers are special objects that can work with the `with ... as ...`: construct. \n",
    "\n",
    "- The object should work like regular files in python (no need to inherit, it would be better to use composition here), but:\n",
    "    - When iterating over an object, we will have to get not a line from a file, but a special `FastaRecord` object. It will store information about the sequence. It is important, not lines, but sequences, in fasta files, a sequence is often split into many lines\n",
    "    - You need to write read_record and read_records methods, which in meaning correspond to `readline()` and `readlines()` in ordinary files, but they should not produce lines, but the `FastaRecord` object (s)\n",
    "- The constructor must take one argument - the path to the file\n",
    "- The class must manage memory efficiently, with the expectation of working with very large files.\n",
    "\n",
    "`FastaRecord` object. It should be a data class with three fields:\n",
    "- `seq` - sequence\n",
    "- `id_` - sequence ID (this is what is in the fast file in the line that starts with > before the first space. For example, >GTD326487.1 Species anonymous 24 chromosome)\n",
    "- `description` - what is left after the ID (For example, >GTD326487.1 Species anonymous 24 chromosome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from dataclasses import dataclass\n",
    "\n",
    "\n",
    "class OpenFasta:\n",
    "    '''A class for reading .fasta files with biological sequences\n",
    "    \n",
    "    Record example: >Seq_ID Description(optional)\n",
    "                    ATGTGGGAGGAAGGTGGGTGAAA\n",
    "             \n",
    "    :atribute str path: path to a .fasta file\n",
    "\n",
    "    :method read_record(): to read one record from .fasta file\n",
    "    :method read_records(): to read multiple records from .fasta file\n",
    "        \n",
    "    :Raises: ValueError if .fasta file does not exist in path        \n",
    "    '''\n",
    "    \n",
    "    \n",
    "    def __init__(self, path):\n",
    "        '''Checks the existance of .fasta file and constructs all the necessary attributes\n",
    "\n",
    "        :param str path: path to a .fasta file\n",
    "        \n",
    "        :raises: ValueError if .fasta file does not exist in path\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        if not os.path.isfile(path):\n",
    "            file_path = path.split(os.path.sep)\n",
    "            raise ValueError(f'No such {file_path[-1]}. Check path to file!')\n",
    "    \n",
    "        self.path = path\n",
    "        \n",
    "    \n",
    "    def __enter__(self):\n",
    "        self.f = open(self.path)\n",
    "        return self\n",
    "    \n",
    "    \n",
    "    def __iter__(self):\n",
    "        return self\n",
    "    \n",
    "    \n",
    "    def __next__(self):\n",
    "        self.seq = []\n",
    "        while True: \n",
    "            line = self.f.readline()\n",
    "            if not line:\n",
    "                if self.seq:\n",
    "                    self.seq = ''.join(self.seq)\n",
    "                    record = FastaRecord(id_=self.id_,\n",
    "                                         description=self.description,\n",
    "                                         seq=self.seq)\n",
    "                    return record\n",
    "                return ''\n",
    "            elif line.startswith('>'):\n",
    "                if self.seq:\n",
    "                    self.seq = ''.join(self.seq)\n",
    "                    record = FastaRecord(id_=self.id_,\n",
    "                                         description=self.description,\n",
    "                                         seq=self.seq)\n",
    "                    self.id_, self.description = self.__handle_id_descr(line)\n",
    "                    return record\n",
    "                self.id_, self.description = self.__handle_id_descr(line)\n",
    "            else:\n",
    "                self.seq.append(line.strip())\n",
    "                \n",
    "                \n",
    "    def __exit__(self, exc_type, exc_value, exc_tb):\n",
    "        self.f.close()\n",
    "        pass\n",
    "               \n",
    "            \n",
    "    def __handle_id_descr(self, line):\n",
    "        '''Handles the first record line with Seq_ID and Description (optional)\n",
    "        \n",
    "        :param str line: record line\n",
    "        :returns: record Seq_ID and record description\n",
    "        :rtype: (str, str)\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        vals = line.strip().split(' ', maxsplit=1)\n",
    "        id_ = vals[0][1:]\n",
    "        if len(vals) == 1:\n",
    "            description = ''\n",
    "        else: \n",
    "            description = vals[1]\n",
    "        return id_, description\n",
    "            \n",
    "                \n",
    "    def read_record(self):\n",
    "        '''Reads fasta record\n",
    "        \n",
    "        :returns: FastaRecord\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        return next(self)\n",
    "    \n",
    "    \n",
    "    def read_records(self):\n",
    "        '''Reads all fasta records in file and appends them to list\n",
    "        \n",
    "        :returns: records \n",
    "        :rtype: list\n",
    "        '''\n",
    "        \n",
    "        \n",
    "        records = []\n",
    "        record = self.read_record()\n",
    "        while record:\n",
    "            records.append(record)\n",
    "            record = self.read_record()\n",
    "        return records\n",
    "              \n",
    "\n",
    "@dataclass\n",
    "class FastaRecord:\n",
    "    '''A dataclass with information about fasta record\n",
    "    :param str id_: record id\n",
    "    :param str description: record description\n",
    "    :param str seq: sequnce of nucleotides or aminoacids\n",
    "    '''\n",
    "    \n",
    "    id_: str\n",
    "    description: str\n",
    "    seq: str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "______________________________\n",
      "FastaRecord form file:\n",
      "FastaRecord(id_='SUP35_Kla_AB039749', description='', seq='ATGTCAGACCAACAAAATCAAGACCAAGGGCAAGGCCAAGGTTACAATCAGTATAACCAATATGGCCAGTACAACCAGTACTACAACCAACAGGGCTATCAAGGCTACAACGGCCAACAAGGTGCTCCTCAAGGCTACCAAGCATATCAAGCTTATGGACAGCAGCCTCAAGGAGCCTACCAGGGCTACAACCCTCAACAAGCTCAAGGCTACCAACCTTACCAGGGCTACAATGCTCAGCAACAAGGTTACAACGCTCAGCAAGGCGGTCACAAC')\n",
      "Seq_ID: SUP35_Kla_AB039749\n",
      "Seq: ATGTCAGACCAACAAAATCAAGACCAAGGGCAAGGCCAAGGTTACAATCAGTATAACCAATATGGCCAGTACAACCAGTACTACAACCAACAGGGCTATCAAGGCTACAACGGCCAACAAGGTGCTCCTCAAGGCTACCAAGCATATCAAGCTTATGGACAGCAGCCTCAAGGAGCCTACCAGGGCTACAACCCTCAACAAGCTCAAGGCTACCAACCTTACCAGGGCTACAATGCTCAGCAACAAGGTTACAACGCTCAGCAAGGCGGTCACAAC\n",
      "<class '__main__.FastaRecord'>\n",
      "______________________________\n",
      "All FastaRecords from file:\n",
      "FastaRecord(id_='SUP35_Kla_AB039749', description='', seq='ATGTCAGACCAACAAAATCAAGACCAAGGGCAAGGCCAAGGTTACAATCAGTATAACCAATATGGCCAGTACAACCAGTACTACAACCAACAGGGCTATCAAGGCTACAACGGCCAACAAGGTGCTCCTCAAGGCTACCAAGCATATCAAGCTTATGGACAGCAGCCTCAAGGAGCCTACCAGGGCTACAACCCTCAACAAGCTCAAGGCTACCAACCTTACCAGGGCTACAATGCTCAGCAACAAGGTTACAACGCTCAGCAAGGCGGTCACAAC')\n",
      "FastaRecord(id_='SUP35_Agos_ATCC_10895_NM_211584', description='', seq='ATGTCGGAGGAAGATCAAATTCAATCGCAAGGCAACGACCAAGGCCAGTCGCAAGCCAAGGATCAAGGTCAAAATCAAGGTCAGGGGCAGCAAAATTTCGGCCAATACTACAACCCAAGTAACTTCCAGAATTACCAGGGGTATGTGCCTCAGGGAGGTTACCAGGCGTATGGCCAACAGGCAGGTGGGTACCAAGGCTATGC')\n",
      "FastaRecord(id_='SUP35_Scer_74-D694_GCA_001578265.1', description='', seq='ATGTCGGATTCAAACCAAGGCAACAATCAGCAAAACTACCAGCAATACAGCCAGAACGGTAACCAACAACAAGGTAACAACAGATACCAAGGTTATCAAGCTTACAATAACAATTTGCAAGGATATCAAGCTGGTT')\n",
      "FastaRecord(id_='SUP35_Spar_A12_Liti', description='', seq='ATGTCGGATTCAAACCAAGGTAACAATCAGCAAAGCTACCAGCAATACGGCCAAAACCCTAACCAACAACAAGGTAACAACAGATACCAAGGTTA')\n",
      "FastaRecord(id_='SUP35_Smik_IFO1815T_30', description='', seq='ATGTCTGATTCAAACCAAGGTAATAATCAGCAAAACTACCAGCAATACAACCAAAACCCCAACCAACAGCAAGGTAACAATAGATACCAAGGTTATCAAGCTTATAATGCTCAAGCGCAACC')\n",
      "FastaRecord(id_='SUP35_Skud_IFO1802T_36', description='', seq='ATGTCAGATCCAAATCAAGGTAACAATCAACAACAATACGGTCAAAATCCTAACCAACAGCAAGGCAATAACAAGTACCAAGGTTATCAAGCTTACAATGCTCAAGCCCAACCTGCAAAC')\n",
      "FastaRecord(id_='SUP35_Sbou_unique28_CM003560', description='', seq='ATGTCGGATTCAAACCAAGGCAACAATCAGCA')\n",
      "______________________________\n",
      "FastaRecord(id_='SUP35_Kla_AB039749', description='', seq='ATGTCAGACCAACAAAATCAAGACCAAGGGCAAGGCCAAGGTTACAATCAGTATAACCAATATGGCCAGTACAACCAGTACTACAACCAACAGGGCTATCAAGGCTACAACGGCCAACAAGGTGCTCCTCAAGGCTACCAAGCATATCAAGCTTATGGACAGCAGCCTCAAGGAGCCTACCAGGGCTACAACCCTCAACAAGCTCAAGGCTACCAACCTTACCAGGGCTACAATGCTCAGCAACAAGGTTACAACGCTCAGCAAGGCGGTCACAAC')\n",
      "FastaRecord(id_='SUP35_Agos_ATCC_10895_NM_211584', description='', seq='ATGTCGGAGGAAGATCAAATTCAATCGCAAGGCAACGACCAAGGCCAGTCGCAAGCCAAGGATCAAGGTCAAAATCAAGGTCAGGGGCAGCAAAATTTCGGCCAATACTACAACCCAAGTAACTTCCAGAATTACCAGGGGTATGTGCCTCAGGGAGGTTACCAGGCGTATGGCCAACAGGCAGGTGGGTACCAAGGCTATGC')\n",
      "FastaRecord(id_='SUP35_Scer_74-D694_GCA_001578265.1', description='', seq='ATGTCGGATTCAAACCAAGGCAACAATCAGCAAAACTACCAGCAATACAGCCAGAACGGTAACCAACAACAAGGTAACAACAGATACCAAGGTTATCAAGCTTACAATAACAATTTGCAAGGATATCAAGCTGGTT')\n",
      "FastaRecord(id_='SUP35_Spar_A12_Liti', description='', seq='ATGTCGGATTCAAACCAAGGTAACAATCAGCAAAGCTACCAGCAATACGGCCAAAACCCTAACCAACAACAAGGTAACAACAGATACCAAGGTTA')\n",
      "FastaRecord(id_='SUP35_Smik_IFO1815T_30', description='', seq='ATGTCTGATTCAAACCAAGGTAATAATCAGCAAAACTACCAGCAATACAACCAAAACCCCAACCAACAGCAAGGTAACAATAGATACCAAGGTTATCAAGCTTATAATGCTCAAGCGCAACC')\n",
      "FastaRecord(id_='SUP35_Skud_IFO1802T_36', description='', seq='ATGTCAGATCCAAATCAAGGTAACAATCAACAACAATACGGTCAAAATCCTAACCAACAGCAAGGCAATAACAAGTACCAAGGTTATCAAGCTTACAATGCTCAAGCCCAACCTGCAAAC')\n",
      "FastaRecord(id_='SUP35_Sbou_unique28_CM003560', description='', seq='ATGTCGGATTCAAACCAAGGCAACAATCAGCA')\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#ex1.Reading one fasta record from file\n",
    "with OpenFasta(os.path.join(\"data\", \"example.fasta\")) as fasta:\n",
    "    record = fasta.read_record()\n",
    "    print('_'*30)\n",
    "    print(f'FastaRecord form file:\\n{record}')\n",
    "    print(f'Seq_ID: {record.id_}\\nSeq: {record.seq}')\n",
    "    print(type(record))\n",
    "\n",
    "#ex2. Reading all fasta records from file\n",
    "with OpenFasta(os.path.join(\"data\", \"example.fasta\")) as fasta:\n",
    "    print('_'*30)\n",
    "    print(f'All FastaRecords from file:')\n",
    "    for record in fasta.read_records():\n",
    "        print(record)\n",
    "\n",
    "#ex3. At the end of the file acts like `readlines()\n",
    "with OpenFasta(os.path.join(\"data\", \"example.fasta\")) as fasta:\n",
    "    print('_'*30)\n",
    "    for _ in range(15):\n",
    "        print(fasta.read_record())  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Wrapping with lru_cache  \n",
    "1. Write a code that will allow you to get all possible (non-unique) genotypes when two organisms are crossed. It can be a function or a class, whichever is more convenient for you. \n",
    "\n",
    "For example, all possible outcomes of crossing \"Aabb\" and \"Aabb\" (non-unique) are\n",
    "```\n",
    "AAbb\n",
    "AAbb\n",
    "AAbb\n",
    "AAbb\n",
    "Aabb\n",
    "Aabb\n",
    "Aabb\n",
    "Aabb\n",
    "Aabb\n",
    "Aabb\n",
    "Aabb\n",
    "Aabb\n",
    "aabb\n",
    "aabb\n",
    "aabb\n",
    "aabb\n",
    "```\n",
    "2. Write a function that calculates the probability of occurrence of a certain genotype (its expected proportion in the offspring). \n",
    "\n",
    "For example,\n",
    "```python\n",
    "get_offspting_genotype_probability(parent1=\"Aabb\", parent2=\"Aabb\", target_genotype=\"Aabb\") # 0.5\n",
    "```\n",
    "\n",
    "3. Write a code that outputs all unique genotypes when crossing `'АаБбввГгДдЕеЖжЗзИиЙйккЛлМмНн'` and `'АаббВвГгДДЕеЖжЗзИиЙйКкЛлМмНН'`, which contain the following combination of alleles `'АаБбВвГгДдЕеЖжЗзИиЙйКкЛл'`\n",
    "\n",
    "4. Write a code that calculates the probability of occurrence of the genotype `'АаБбввГгДдЕеЖжЗзИиЙйккЛлМмНн'` when crossing `АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНн` and `АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНн`\n",
    "\n",
    "\n",
    "Important notes:\n",
    "\n",
    "1. The order of the alleles in the case of heterozygotes should always be as follows: first a capital letter, then a small one (AaBb is acceptable, but aAbB should not be)\n",
    "2. Subtasks 3 and 4 can be computationally intensive (up to 15+ minutes depending on the hardware), so make sure you test your code well on small data before doing these tasks. If your code runs longer than 20 minutes, then most likely your solution is not optimal, try to optimize something. If the optimal solution does not work at all, then try to remove the last gene from the input data in all tasks (this should reduce the execution time by about 4 times), but 2 points will be deducted for such a solution\n",
    "3. Although it is possible to solve subtasks 2, 3 and 4 mathematically without resorting to obtaining all possible genotypes directly, it is the brute-force version of the algorithm that is required from you."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import lru_cache\n",
    "from datetime import datetime\n",
    "\n",
    "\n",
    "# Subtasks 1 and 2\n",
    "def calc_runtime(func):\n",
    "    '''Decorator for calculating of function runtime'''\n",
    "    \n",
    "    \n",
    "    def inner_function(*args, **kwargs):\n",
    "        start = datetime.now()\n",
    "        result = func(*args, **kwargs)\n",
    "        stop = datetime.now()\n",
    "        print(stop - start)\n",
    "        return result\n",
    "    return inner_function  \n",
    "\n",
    "\n",
    "def parse_genotype(genotype):\n",
    "    '''Parses genotypes to obtain alleles\n",
    "    \n",
    "    :param str genotype: genotype, example: 'Aabb'\n",
    "    :returns: list of alleles\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    return [genotype[i:i+2] for i in range(0, len(genotype), 2)]\n",
    "\n",
    "\n",
    "def calc_gamets(*args):\n",
    "    '''Creates sets with all possible gamets for genotype\n",
    "    \n",
    "    :param args: alleles returned by parse_genotype()\n",
    "    :returns: generator with all possible gamets for genotype\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    pools = [tuple(pool) for pool in args]\n",
    "    result = [[]]\n",
    "    for pool in pools:\n",
    "        result = [x+[y] for x in result for y in pool]\n",
    "    for prod in result:\n",
    "        yield tuple(prod)\n",
    "        \n",
    "\n",
    "@lru_cache(maxsize=None)\n",
    "def do_genotype(*gamets):\n",
    "    '''Makes genotype from gamets\n",
    "    \n",
    "    :param tuples gamets: two sets of gamets\n",
    "    :returns str: resulted genotype\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    genotype = ''.join([''.join(sorted(gene)) for gene in zip(*gamets)])\n",
    "    return genotype\n",
    "\n",
    "        \n",
    "def calc_genotypes(gamets1, gamets2, k=1):\n",
    "    '''Creates all possible genotypes from two sets of gamets\n",
    "    \n",
    "    :param tuple gamets1: set of gamets for the first genotype\n",
    "    :param tuple gamets2: set of gamets for the second genotype\n",
    "    :param int k: number of genes devided by two\n",
    "    :retutns: generator with possible genotypes\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    for gamet1 in gamets1:\n",
    "        for gamet2 in gamets2:\n",
    "            geno_l = do_genotype(gamet1[:k], gamet2[:k])\n",
    "            geno_r = do_genotype(gamet1[k:], gamet2[k:])\n",
    "            yield geno_l + geno_r   \n",
    "\n",
    "            \n",
    "def all_genotypes(p1, p2):\n",
    "    '''Creates all possible offspring genotypes for given parental genotypes\n",
    "    \n",
    "    :param str p1: parent1 genotype\n",
    "    :param str p2: parent2 genotype\n",
    "    :returns: None\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    gamets1 = tuple(calc_gamets(*parse_genotype(p1)))\n",
    "    gamets2 = tuple(calc_gamets(*parse_genotype(p2)))\n",
    "    [print(genotype) for genotype in calc_genotypes(gamets1, gamets2)]\n",
    "    \n",
    "\n",
    "@calc_runtime\n",
    "def genotype_prob(p1, p2, target):\n",
    "    '''Calculates the probability of a certain genotype (its expected share in the offspring)\n",
    "    \n",
    "    :param str p1: parent1 genotype\n",
    "    :param str p2: parent2 genotype\n",
    "    :param str target: target offspring genotype for its probability calculation\n",
    "    :returns: probability of a certain genotype\n",
    "    :rtype: float\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    gamets1 = tuple(calc_gamets(*parse_genotype(p1)))\n",
    "    gamets2 = tuple(calc_gamets(*parse_genotype(p2)))\n",
    "    \n",
    "    total_count = target_count = 0\n",
    "    k = len(p1) // 4\n",
    "    genotypes = calc_genotypes(gamets1, gamets2, k)\n",
    "    for genotype in genotypes:\n",
    "        total_count += 1\n",
    "        if target == genotype:\n",
    "            target_count +=1\n",
    "    return target_count/total_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAbb\n",
      "AAbb\n",
      "Aabb\n",
      "Aabb\n",
      "AAbb\n",
      "AAbb\n",
      "Aabb\n",
      "Aabb\n",
      "Aabb\n",
      "Aabb\n",
      "aabb\n",
      "aabb\n",
      "Aabb\n",
      "Aabb\n",
      "aabb\n",
      "aabb\n"
     ]
    }
   ],
   "source": [
    "# example for Subtask 1\n",
    "parent1 = 'Aabb'\n",
    "parent2 = 'Aabb'\n",
    "all_genotypes(parent1, parent2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.000496\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.5"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# example for Subtask 2\n",
    "parent1 = 'Aabb'\n",
    "parent2 = 'Aabb'\n",
    "target = 'Aabb'\n",
    "genotype_prob(parent1, parent2, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subtask 3\n",
    "@calc_runtime\n",
    "def unique_with(p1, p2, target):\n",
    "    '''Finds all unique genotypes with certain alleles combination\n",
    "    \n",
    "    Parameters:\n",
    "        p1 (str): parent1 genotype\n",
    "        p2 (str): parent2 genotype\n",
    "        target (str): target offspring genotype\n",
    "        \n",
    "    Returns: None\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    gamets1 = tuple(calc_gamets(*parse_genotype(p1)))\n",
    "    gamets2 = tuple(calc_gamets(*parse_genotype(p2)))\n",
    "    \n",
    "    unique = []\n",
    "    for genotype in calc_genotypes(gamets1, gamets2):\n",
    "        if (genotype.find(target) != -1) and (genotype not in unique):\n",
    "            unique += [genotype]\n",
    "            print(genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "АаБбВвГгДдЕеЖжЗзИиЙйКкЛлММНН\n",
      "АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНН\n",
      "АаБбВвГгДдЕеЖжЗзИиЙйКкЛлММНн\n",
      "АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНн\n",
      "АаБбВвГгДдЕеЖжЗзИиЙйКкЛлммНН\n",
      "АаБбВвГгДдЕеЖжЗзИиЙйКкЛлммНн\n",
      "0:08:33.601783\n"
     ]
    }
   ],
   "source": [
    "# example for Subtask 3\n",
    "parent1 = 'АаБбввГгДдЕеЖжЗзИиЙйккЛлМмНн'\n",
    "parent2 = 'АаббВвГгДДЕеЖжЗзИиЙйКкЛлМмНН'\n",
    "target = 'АаБбВвГгДдЕеЖжЗзИиЙйКкЛл'\n",
    "unique_with(parent1, parent2, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:06:41.638304\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1.52587890625e-05"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# example for Subtask4\n",
    "parent1 = 'АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНн'\n",
    "parent2 = 'АаБбВвГгДдЕеЖжЗзИиЙйКкЛлМмНн'\n",
    "target = 'АаБбввГгДдЕеЖжЗзИиЙйккЛлМмНн'\n",
    "genotype_prob(parent1, parent2, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29.7"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "540*0.055"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
